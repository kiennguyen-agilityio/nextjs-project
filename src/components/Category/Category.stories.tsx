import type { Meta, StoryObj } from '@storybook/react';

// components
import Category from '@/components/Category';

// mocks
import { roles } from '@/mocks';

const meta = {
  title: 'Components/Category',
  component: Category,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Category>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    roles: roles,
  },
};
