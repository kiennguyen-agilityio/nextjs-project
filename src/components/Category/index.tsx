'use client';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';

// mocks
import { listCategory } from '@/mocks';

// components
import Dropdown from '@/components/common/Dropdown';
import { RoleModel } from '@/models/RoleModel';

interface CategoryProps {
  roles: RoleModel[];
}

const Category = ({ roles }: CategoryProps) => {
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const router = useRouter();

  const createFilterURL = (name: string, value: string) => {
    const params = new URLSearchParams(searchParams);

    if (name === 'userRole') {
      params.set(name, value);
    } else {
      params.set('sortBy', name);
      params.set('order', value);
    }

    router.push(`${pathname}?${params.toString()}`);
  };

  return (
    <thead>
      <tr className="flex items-center justify-between pl-0 md:pl-0 pr-8 md:pr-12 border-b border-gray-200">
        {listCategory.map(({ id, label, options = [], name }) => (
          <th key={id} className="text-left mb-4 sm:mb-0">
            <Dropdown
              value={id}
              label={label}
              options={
                label === 'Role'
                  ? roles.map((role) => ({
                      id: role.id.toString(),
                      label: role.name,
                    }))
                  : options.map((option) => ({
                      id: option.value,
                      ...option,
                    }))
              }
              onChange={(value) => {
                // value name

                createFilterURL(name, value);
              }}
            />
          </th>
        ))}
        <th className="text-left flex items-center mb-4 sm:mb-0" />
      </tr>
    </thead>
  );
};

export default Category;
