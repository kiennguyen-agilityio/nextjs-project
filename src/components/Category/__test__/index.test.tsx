import { render } from '@testing-library/react';

// components
import Category from '@/components/Category';

test('renders Category component with dropdown buttons', () => {
  const { asFragment } = render(
    <table>
      <Category />
    </table>,
  );
  expect(asFragment()).toMatchSnapshot();
});
