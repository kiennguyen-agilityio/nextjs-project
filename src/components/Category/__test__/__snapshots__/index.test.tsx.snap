// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Category component with dropdown buttons 1`] = `
<DocumentFragment>
  <select
    class="inline-flex justify-center w-full rounded-3xl px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-200 focus:outline-none max-w-32 undefined"
  >
    <option
      value=""
    >
      Email
    </option>
    <option
      value="email1"
    >
      email1@example.com
    </option>
    <option
      value="email2"
    >
      email2@example.com
    </option>
  </select>
  <select
    class="inline-flex justify-center w-full rounded-3xl px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-200 focus:outline-none max-w-32 undefined"
  >
    <option
      value=""
    >
      Full Name
    </option>
    <option
      value="john_doe"
    >
      John Doe
    </option>
    <option
      value="jane_doe"
    >
      Jane Doe
    </option>
  </select>
  <select
    class="inline-flex justify-center w-full rounded-3xl px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-200 focus:outline-none max-w-32 undefined"
  >
    <option
      value=""
    >
      Role
    </option>
  </select>
  <select
    class="inline-flex justify-center w-full rounded-3xl px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-200 focus:outline-none max-w-32 undefined"
  >
    <option
      value=""
    >
      Joined
    </option>
    <option
      value="2021"
    >
      2021
    </option>
    <option
      value="2022"
    >
      2022
    </option>
  </select>
</DocumentFragment>
`;
